### DISCUSSION TITLE
Could Mapster notify about an unmapped property of the source object?

### DISCUSSION THREAD

**Discussion by plenartowicz** [MapsterMapper/Mapster#739] [Q&A]
I love Mapster because of I don't have to write any configuration. It just works. In most cases, this is an advantage.
But when renaming properties this sometimes becomes a drawback.

For instance:
```cs
public class Source
{
    public int Id { get; set; }
    public string Name { get; set; }
    public decimal Value { get; set; }
}

public class Destination
{
    public int Id { get; set; }
    public string Name { get; set; }
    public decimal Value { get; set; }
}

var s = new Source();
var d = s.Adapt<Destination>();  
```

When I rename `Name` property of class `Source`, after adapting `d.Name` will be `null`. 
```cs
public class Source
{
    public int Id { get; set; }
    public string Description { get; set; }  // I rename this
    public decimal Value { get; set; }
}

public class Destination
{
    public int Id { get; set; }
    public string Name { get; set; }.  // I forget/don't know to rename this
    public decimal Value { get; set; }
}

var s = new Source();
var d = s.Adapt<Destination>();    // here, d.Name will be null
```
There is no configuration so Mapster couldn't adapt `Description` to `Name. But I would like to know about this. 
So the question is: Is there any way to Mapster notifies about this or provides configuration for this behavior?

I couldn't find any setting which helps me.

**Comment by stagep** [comment]
Are you asking if there is a way to configure Mapster so that Name is mapped to Description, or are you asking if there is some way to be notified that the mapping (without any configuration) is creating a null value?

Do you think that a runtime exception is really the best way to handle this rather than having unit tests that would catch this?

**Reply by plenartowicz** [reply]
I would like to know when Mapster couldn't find way to map (in my case) `Description` property to other one. It could be a simple log, `Debug.WriteLine` (just like WPF notify about errors during binding process) or a setting to add custom behavior.

I agree with you the runtime exception is not the best way. On the other hand unit tests for all mapping could be a little bit difficult. 

Honestly, I just want to know if it's even possible.

**Comment by stagep** [comment]
The issue would seem to be that you have various types and you want to map from type A to type B, and from type C to type D and you want to make sure that the types you want to map to have the same properties as the type you are mapping from. So this doesn't even need to involve Mapster.  With xUnit you could do the following:

```
    [Theory]
    [InlineData(typeof(Source), typeof(Destination))]
    [InlineData(typeof(Source2), typeof(Destination2))]
    public void TypesHaveSameProperties(Type type1, Type type2)
    {
        var properties1 = type1.GetProperties(BindingFlags.Public | BindingFlags.Instance);
        var properties2 = type2.GetProperties(BindingFlags.Public | BindingFlags.Instance);

        Assert.Equal(properties1.Length, properties2.Length);

        foreach (var prop1 in properties1)
        {
            var prop2 = properties2.FirstOrDefault(p => p.Name == prop1.Name);
            Assert.NotNull(prop2);
            Assert.Equal(prop1.PropertyType, prop2.PropertyType);
        }
    }
```

**Reply by plenartowicz** [reply]
Maybe in my simple example unit tests could help but they are not what I want. 

I'm looking for a setting so that Mapster could notify about an unmapped property of the source object.

Of course, I realize that there may be no such setting but that's why I'm asking this question here.
